File: haskell/glue/src/Glue/Lib/List/Map.hs
module Glue.Lib.List.Map where

import Glue.Eval (Eval, apply, throwError)
import Glue.Eval.Exception
import Glue.IR (IR (..))

map :: IR Eval
map = NativeFunc mapImpl

mapImpl :: IR Eval -> Eval (IR Eval)
mapImpl funcIR = pure $ NativeFunc (mapOver funcIR)

mapOver :: IR Eval -> IR Eval -> Eval (IR Eval)
mapOver func list = case list of
    List xs -> do
        -- Apply the function to each element by evaluating a list [func, x]
        results <- mapM (\x -> apply func [x]) xs
        pure $ List results
    _ -> throwError $ wrongArgumentType ["function", "list"]

File: dart/glue/lib/src/lib/list/map.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Map function - applies a function to each element of a list
/// Mirrors Haskell Glue.Lib.List.Map.map exactly
Ir map = IrNativeFunc(mapImpl);

/// Map function implementation
/// Mirrors Haskell Glue.Lib.List.Map.mapImpl exactly
Eval<Ir> mapImpl(Ir funcIr) {
  return Eval.pure(IrNativeFunc(mapOver(funcIr)));
}

/// Helper function for list argument
/// Mirrors Haskell Glue.Lib.List.Map.mapOver exactly
Eval<Ir> Function(Ir) mapOver(Ir funcIr) {
  return (Ir listIr) {
    return switch (listIr) {
      IrList(elements: final elements) => sequenceAll(
        elements.map((element) => apply(funcIr, [element])).toList(),
      ).map((results) => IrList(results)),
      _ => throwError(wrongArgumentType(['function', 'list'])),
    };
  };
}
